### ISSUE TITLE
empty command line arguments filtered out on windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3876]
|<img alt="anton's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| anton reported|
|-|-|
|Trac ID|trac#3876|
|Type|defect|
|Created|2009-06-10 08:46:43Z|
|Branch|https://github.com/twisted/twisted/tree/empty-win-args-3876|

If I run something like ```getProcessOutput('some.exe', ('foo', '', 'bar'))``` on windows, the empty command line argument is filtered out and the command line that is actually run is ```'some.exe foo  bar'``` (two spaces between foo and bar).

This can be fixed by improving the argument quoting in twisted.python.win32.cmdLineQuote:
```
--- twisted/python/win32.py     2009-06-10 11:01:54.753700000 +0300
+++ twisted/python/win32.py.orig        2009-06-10 11:01:22.457700000 +0300
@@ -75,7 +75,7 @@

     @return: a quoted string.
     """
-    quote = ((" " in s) or ("\t" in s) or ('"' in s) or s == '') and '"' or ''
+    quote = ((" " in s) or ("\t" in s) or ('"' in s)) and '"' or ''
     return quote + _cmdLineQuoteRe2.sub(r"\1\1", _cmdLineQuoteRe.sub(r'\1\1\\"', s)) + quote

 def quoteArguments(arguments):

```

The above would lead to the example command line becoming ```'some.exe foo "" bar'``` and ```""``` would be correctly unqoted as an empty argument.

<details><summary>Searchable metadata</summary>

```
trac-id__3876 3876
type__defect defect
reporter__anton anton
priority__normal normal
milestone__ 
branch__branches_empty_win_args_3876 branches/empty-win-args-3876
branch_author__khorn khorn
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1244623603000000 1244623603000000
changetime__1266368664000000 1266368664000000
version__None None
owner__ 
cc__khorn cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to anton|
|-|-|

Thanks for the fix!

Could you by any chance provide a patch to the unit tests as well?  I'd love to help you at our next sprint, if you're not familiar with this...

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)|@khorn set owner to @khorn|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_3"></a>|@khorn commented|
|-|-|

(In [28183]) Branching to 'empty-win-args-3876'

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_4"></a>|@khorn commented|
|-|-|

(In [28184]) patch applied with tests)

Refs: [#3876](https://github.com/twisted/twisted/issues/3876)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)|@khorn removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess set owner to @khorn|
|-|-|

Thanks for the patch and test cases!

win32.py
 * copyright bump
 * I don't know if the the # -*- test-case-name: twisted.test.test_foo -*- comments get generated by hand or something else, but win32.py should have one
 * In cmdLineQuote, use @type and @rtype markup to describe the types for @param and@return. Also, I don't know if this actually affects the generated documentation, but markup that spans more than one line should be indented (eg the ivars in class _ErrorFormatter).

test_win32.py
 * Use plural forms of asserts (the coding standard in trunk recently got updated to reflect this after some months of confusion)
 * Use two blank lines between class-level functions
 * test_argWithSpaces and test_argWithoutSpaces should swap names :)
 * It looks like you can set skip at the class level instead of skipping individual tests (eg. twisted/conch/test/test_transport.py
 * there's a bunch of trailing whitespace in the file

topfile
 * \ No newline at end of file

[builds](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/empty-win-args-3876) look good.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_8"></a>|@khorn commented|
|-|-|

(In [28202]) Address review comments.

Refs [#3876](https://github.com/twisted/twisted/issues/3876)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_9"></a>|@khorn removed owner|
|-|-|

Addressed review comments in changeset listed above.  Sending back to review.

Note to reviewer: It was pointed out on IRC that there was really no reason to skip these tests on non-Windows platforms.  Even though the code under test is intended for use on Windows it should run just about anywhere, as should the tests.  This made sense to me, so I removed the skips from the tests.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess set owner to @khorn|
|-|-|

The win32.py copyright thinks it's 20010 :)

Other than that, looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_12"></a>|@khorn commented|
|-|-|

(In [28228]) Fix a typo

Refs: [#3876](https://github.com/twisted/twisted/issues/3876)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_13"></a>|@khorn commented|
|-|-|

Replying to [khorn](#note_12):
> (In [28228]) Fix a typo
>
> Refs: [#3876](https://github.com/twisted/twisted/issues/3876)

Ignore that...I put in the wrong ticket number...dagnabbit.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_14"></a>|@khorn commented|
|-|-|

No I put in the right ticket number...that belongs here after all...I just went crazy for a minute...<sigh>

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_15"></a>|@khorn set status to `closed`|
|-|-|

(In [28230]) Merge empty-win-args-3876: Parse empty command line arguments on Windows

twisted.python.win32.cmdLineQuote now correctly quotes empty strings arguments

Author: anton, khorn
Reviewer: jesstess
Fixes: [#3876](https://github.com/twisted/twisted/issues/3876)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
